lib/exqlite/rw_connection.ex

```diff
@@ -0,0 +1,274 @@
+defmodule Exqlite.RWConnection do
+  @moduledoc """
+  RWConnection owns two SQLite3 open databases. One is used for serialized writes and the other is used for concurrent reads.
+  When a write finishes, the "reads" database is drained of statements to get out of implicit transaction and see new data.
+  """
+
+  use GenServer
+  alias Exqlite.Sqlite3
+
+  @type t :: GenServer.server()
+
+  @doc """
+  Starts a connection process.
+
+  ## Options
+
+    * `:database` (required) - the database process to connect to
+
+    * `:process_options` - the options to be given to the underlying
+      process. See `GenServer.start_link/3` for all options
+
+  ## Examples
+
+      #{__MODULE__}.start_link(
+        database: ":memory:",
+        process_options: [name: MyApp.Conn]
+      )
+
+  In your supervision tree it would be started like this:
+
+      children = [
+        {#{__MODULE__},
+         database: ":memory:",
+         process_options: [name: MyApp.Conn]}
+      ]
+
+  """
+  def start_link(options) do
+    {process_options, options} = Keyword.pop(options, :process_options, [])
+    GenServer.start_link(__MODULE__, options, process_options)
+  end
+
+  def stop(conn), do: GenServer.stop(conn)
+
+  @doc """
+  Runs the given `query` with `params` while exclusively locking the connection.
+
+  ## Examples
+
+      query(MyApp.Conn, "create table users(name)")
+      query(conn, "insert into users(names) values (?), (?)", ["bim", "bom"])
+
+  """
+  def query(conn, query, params \\ []) do
+    command(conn, :lock, {:query, query, params}, &result/2)
+  end
+
+  @doc """
+  Runs the given `query` with `params` without locking the connection.
+  Should only be used for reading.
+
+  ## Examples
+
+      read_query(MyApp.Conn, "select * from users")
+      read_query(conn, "select * from users where name = ?", ["bimbom"])
+
+  """
+  def read_query(conn, query, params \\ []) do
+    command(conn, :read, {:query, query, params}, &result/2)
+  end
+
+  defp command(conn, type, command, fun) do
+    case GenServer.call(conn, {type, command}, :infinity) do
+      {:ok, db, undo_ref, statement_ref} ->
+        try do
+          fun.(db, statement_ref)
+        after
+          GenServer.cast(conn, {:undo, type, undo_ref})
+        end
+
+      {:error, reason} ->
+        {:error, error_to_exception(reason)}
+    end
+  end
+
+  defp result(db, statement_ref) do
+    with {:ok, columns} <- Sqlite3.columns(db, statement_ref),
+         {:ok, rows} <- Sqlite3.fetch_all(db, statement_ref, _chunk_size = 100),
+         do: {:ok, %Exqlite.Result{columns: columns, rows: rows}}
+  end
+
+  @impl true
+  def init(options) do
+    database = Keyword.fetch!(options, :database)
+
+    if database in [:memory, ":memory:"] do
+      raise "#{database} is not supported in #{__MODULE__}"
+    end
```

Comment:
I think it could be supported. SQLite has internal locking mechanisms that should handle two concurrent reads and writes in memory no problem. 

https://www.sqlite.org/c3ref/threadsafe.html

We have it compiled with

```
CFLAGS += -DSQLITE_THREADSAFE=1
```

Reply:
I thought the databases would be in different places in memory. So inserting in one wouldn't make the data visible in the other.

Reply:
They will be in different places yes. I forgot to send along that I think we could support an in memory database by setting both the reader and writer db instances to be the same DB handles.

Reply:
Ah, right. I'll try that.

Reply:
I think you'd just need to track if it was opened as in memory so you know to close the db once and then dereference the old handle.